src/main/scala/lms/transformation/distributed_tensor/distributedTensorUnary.scala

```diff
@@ -45,6 +63,30 @@ trait FixedSizeDistributedTensorUnaryTypeLess extends FixedSizeDistributedTensor
         (() => {
           Accumulate(gradMap(a), Negate(gradMap(s), anno), anno); ()
         }) +=: backwardNodes
+     
+    case Node(s, "tensor_invert", tt::Backend.Const(anno:Anno)::(a:Backend.Sym)::_, _) =>
+        implicit val pos = Adapter.oldSourceMap(s)
+        forwardNodes += node
+
+        (() => {
+          Accumulate(gradMap(a), Invert(gradMap(s), anno), anno); ()
+        }) +=: backwardNodes
+    
+    case Node(s, "tensor_tanh", tt::Backend.Const(anno:Anno)::(a:Backend.Sym)::_, _) =>
+        implicit val pos = Adapter.oldSourceMap(s)
+        forwardNodes += node
+
+        (() => {
+          Accumulate(gradMap(a), Tanh(gradMap(s), anno), anno); ()
+        }) +=: backwardNodes
+    
+    case Node(s, "tensor_relu", tt::Backend.Const(anno:Anno)::(a:Backend.Sym)::_, _) =>
+        implicit val pos = Adapter.oldSourceMap(s)
+        forwardNodes += node
+
+        (() => {
+          Accumulate(gradMap(a), Relu(gradMap(s), anno), anno); ()
```

Comment:
we don't need to worry about derivative of Relu. However, Relu actually behaves as a mask (only positive values are used after Relu). As a result, the gradient (dz/dy) should also be filtered by the same mask before adding to (dz/dx).
so I think we might need a ReluGrad node that takes the input or output of Relu (as a mask), the gradMap(s), and return a masked gradMap(s). such as
Accumulate(gradMap(a), ReluGrad(transform(a or s), gradMap(s), anno), anno); ()
where the computation of ReluGrad is similar to Relu, but it masks the gradMap(s) argument.

Reply:
I added the ReluGrad node. I am not sure if I fully understood how the gradient of Relu was supposed to work so I did a little extra research and found this: https://blog.paperspace.com/vanishing-gradients-activation-function/.
I essentially implemented ReluGrad as x > 0 => 1, x < 0 => 0